From d2347b84e82870632452731b792a9e9e009a3b18 Mon Sep 17 00:00:00 2001
From: William Horvath <william@horvath.blog>
Date: Mon, 9 Sep 2024 00:42:17 -0700
Subject: [PATCH] HACK: combase: Retry InternalIsProcessInitialized() for 200ms
 to see if the process eventually initializes.

Trying to workaround a race condition in starting .NET applications (e.g. osu!)
---
 dlls/combase/combase.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/dlls/combase/combase.c b/dlls/combase/combase.c
index 9a68e92ff42..3c2ab23fe5d 100644
--- a/dlls/combase/combase.c
+++ b/dlls/combase/combase.c
@@ -2653,10 +2653,17 @@ HRESULT WINAPI CoGetContextToken(ULONG_PTR *token)
 {
     struct tlsdata *tlsdata;
     HRESULT hr;
+    DWORD retries;
 
     TRACE("%p\n", token);
 
-    if (!InternalIsProcessInitialized())
+    for (retries = 40; !InternalIsProcessInitialized() && retries > 0; --retries)
+    {
+        LARGE_INTEGER timeout = {.QuadPart = -5 * (ULONGLONG)10000}; /* sleep for 5ms and try again */
+        NtDelayExecution( FALSE, &timeout );
+    }
+
+    if (!retries)
     {
         ERR("apartment not initialised\n");
         return CO_E_NOTINITIALIZED;
-- 
2.46.0

